hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/utils/RationalizedSigMeta.java

```diff
@@ -146,33 +153,60 @@ public void revokeCryptoSigsFrom(final JKey key) {
-        try {
-            if (!payerReqSig.hasHollowKey()) return;
-        } catch (NullPointerException npe) {
-            log.warn("payerReqSig not expected to be null", npe);
+    // add javadocs
+    public void replaceHollowKeysIfNeeded() {
+        if (this == NONE_AVAIL || rationalizedSigs == null) {
             return;
         }
 
-        final var targetEvmAddress = payerReqSig.getHollowKey().getEvmAddress();
+        final var hollowKeysInOtherReqKeysIndex = createHollowKeyIndex();
         for (final var sig : rationalizedSigs) {
-            // maybe do the hashing of the public key a better way... not coupling to Besu classes?
-            final var publicKeyHashed =
-                    Hash.hash(Bytes.of(sig.getExpandedPublicKey())).toArrayUnsafe();
-            if (Arrays.equals(
-                    targetEvmAddress,
-                    0,
-                    targetEvmAddress.length,
-                    publicKeyHashed,
-                    publicKeyHashed.length - 20,
-                    publicKeyHashed.length)) {
-
-                payerReqSig =
-                        new JECDSASecp256k1Key(
-                                MiscCryptoUtils.compressSecp256k1(sig.getExpandedPublicKey()));
-                replacedHollowKey = true;
+            if (sig.getSignatureType() != SignatureType.ECDSA_SECP256K1) {
+                continue;
+            }
+            final var publicKeyHash =
+                    Hash.hash(Bytes.of(sig.getExpandedPublicKeyDirect())).toArrayUnsafe();
+            final var evmAddress =
+                    Bytes.of(
+                            Arrays.copyOfRange(
+                                    publicKeyHash,
+                                    publicKeyHash.length - EVM_ADDRESS_SIZE,
+                                    publicKeyHash.length));
+            final var key =
+                    new JECDSASecp256k1Key(
+                            MiscCryptoUtils.compressSecp256k1(sig.getExpandedPublicKey()));
+            // replace payer JHollowKey if needed and matched
+            final var evmAddressBytes = evmAddress.toArrayUnsafe();
+            if (payerReqSig.hasHollowKey()
+                    && Arrays.equals(payerReqSig.getHollowKey().getEvmAddress(), evmAddressBytes)) {
+                payerReqSig = key;
+            }
+            // replace any other req JHollowKey if needed and matched
+            if (!hollowKeysInOtherReqKeysIndex.isEmpty()
+                    && hollowKeysInOtherReqKeysIndex.containsKey(evmAddress)) {
+                othersReqSigs.set(hollowKeysInOtherReqKeysIndex.get(evmAddress), key);
+                hollowKeysInOtherReqKeysIndex.remove(evmAddress);
+            }
+            activeEvmAddressesToECDSAKeys.put(evmAddressBytes, key);
+        }
+    }
+
+    /*
+       Construct an index to facilitate checking of presence of JHollowKeys
+       in the otherReqSig list and tracking of the indices
+       DISCUSS --- is this 'optimization' worth it?
```

Comment:
Given that most transactions do not have a very large `otherReqSig` list, I'm not sure whether this is worth it or even better... It certainly doesn't make it easier to understand IMO

Maybe iterating directly through the `otherReqSig` list in `replaceHollowKeysIfNeeded` is OK?